haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,215 @@
+                and negative document.
+        """
+        question_pos_doc_neg_doc: List[Dict[str, str]] = []
+
+        for i in range(0, len(question_doc_pairs), batch_size):
+            # question in batches to minimize network latency
+            i_end = min(i + batch_size, len(question_doc_pairs))
+            queries = [e["question"] for e in question_doc_pairs[i:i_end]]
+            pos_docs = [e["document"] for e in question_doc_pairs[i:i_end]]
+
+            # TODO change this to batched queries once they are implemented in retriever
+            # https://github.com/deepset-ai/haystack/issues/1239
+            retriever_results = self.retriever.retrieve(query=queries[0], top_k=self.top_k)
+
+            # iterate through queries and find negatives
+            for question, pos_doc, query_res in zip(queries, pos_docs, [retriever_results]):
+                top_results = query_res
+                random.shuffle(top_results)
+                for doc_item in top_results:
+                    neg_doc = doc_item.content
+                    if neg_doc != pos_doc:
+                        question_pos_doc_neg_doc.append({"question": question, "pos_doc": pos_doc, "neg_doc": neg_doc})
+                        break
+        return question_pos_doc_neg_doc
+
+    def pseudo_labeling(self, mined_negatives: List[Dict[str, str]]) -> List[Dict]:
+        """
+        Given a list of mined negatives, predict the score margin between the positive and negative document using
+        the cross encoder.
+
+        The function returns a list of examples, where each example is a dictionary with the following keys:
+
+        * question: the question string
+        * pos_doc: the positive document string
+        * neg_doc: the negative document string
+        * score: the score margin
+
+        :param mined_negatives: List[Dict[str, str]] = List of mined negatives
+        :type mined_negatives: List[Dict[str, str]]
+        :return: A list of dictionaries, each of which has the following keys:
+            - question: The question string
+            - pos_doc: The positive document string
+            - neg_doc: The negative document string
+            - score: The score margin
+        """
+        examples: List[Dict] = []
+        for item in mined_negatives:
+            scores = self.cross_encoder.predict(
+                [[item["question"], item["pos_doc"]], [item["question"], item["neg_doc"]]]
+            )
+            score_margin = scores[0] - scores[1]
+            examples.append(
+                {
+                    "question": item["question"],
+                    "pos_doc": item["pos_doc"],
+                    "neg_doc": item["neg_doc"],
+                    "score": score_margin,
+                }
+            )
+        return examples
+
+    def run(self, documents: List[Document]) -> Tuple[dict, str]:  # type: ignore
```

Comment:
Maybe it makes sense to rename this to `generate_pseudo_labels` and just return the `pseudo_labels` and have a new `run` method that simply calls this `generate_pseudo_labels` and returns the Tuple of Dict and str. This would make the use of this node outside of a Pipeline more obvious.

Reply:
+1

Reply:
Resolved